File: src/house_robber_ii_213/solution.ts
function rob(nums: number[]): number {
  const n = nums.length;

  if (n === 1) {
    return nums[0];
  }

  if (n === 2) {
    return Math.max(nums[0], nums[1]);
  }

  return Math.max(robHelper(nums.slice(1)), robHelper(nums.slice(0, n - 1)));
}

function robHelper(nums: number[]): number {
  let secondLast = nums[0];
  let last = Math.max(nums[0], nums[1]);

  for (let i = 2; i < nums.length; i++) {
    const current = Math.max(last, secondLast + nums[i]);
    secondLast = last;
    last = current;
  }

  return last;
}

export function test(): void {
  console.log(rob([1, 2, 3, 1]));
  console.log(rob([2, 7, 9, 3, 1]));
}

File: src/house_robber_ii_213/solution.rs
use std::cmp::max;

pub struct SolutionDp;

impl SolutionDp {
    pub fn rob(nums: Vec<i32>) -> i32 {
        let n = nums.len();

        if n == 0 {
            return 0;
        } else if n == 1 {
            return nums[0];
        }

        let mut dp = vec![0; n + 1];

        dp[0] = nums[0];
        dp[1] = max(nums[0], nums[1]);

        for i in 2..n {
            dp[i] = max(dp[i - 1], dp[i - 2] + nums[i]);
        }

        dp[n - 1]
    }
}

pub fn test() {
    println!("{}", SolutionDp::rob(vec![1, 2, 3, 1]));
    println!("{}", SolutionDp::rob(vec![2, 7, 9, 3, 1]));
    println!("{}", SolutionDp::rob(vec![]));
}
